pkg/jobs/jobsprotectedts/jobs_protected_ts_manager.go

```diff
@@ -0,0 +1,225 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package jobsprotectedts
+
+import (
+	"context"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/config"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+)
+
+// timedProtectTimeStampGCPct wait a percentage of the GC time before
+// creating a protected timestamp record.
+const timedProtectTimeStampGCPct = 0.8
+
+// Manager manages protected timestamps installed for jobs, which will
+// install protected timestamps after a certain percentage of the GC interval
+// is hit.
+type Manager struct {
+	db                  *kv.DB
+	codec               keys.SQLCodec
+	protectedTSProvider protectedts.Provider
+	systemConfig        config.SystemConfigProvider
+	jr                  *jobs.Registry
+}
+
+// Cleaner cleans up the protected timestamp record for the job or cancels
+// the installation.
+type Cleaner func(ctx context.Context) error
+
+func setProtectedTSOnJob(details jobspb.Details, u *uuid.UUID) jobspb.Details {
+	switch v := details.(type) {
+	case jobspb.NewSchemaChangeDetails:
+		v.ProtectedTimestampRecord = u
+		return v
+	case jobspb.SchemaChangeDetails:
+		v.ProtectedTimestampRecord = u
+		return v
+	default:
+		panic("not supported")
+	}
+}
+
+func getProtectedTSOnJob(details jobspb.Details) *uuid.UUID {
+	switch v := details.(type) {
+	case jobspb.NewSchemaChangeDetails:
+		return v.ProtectedTimestampRecord
+	case jobspb.SchemaChangeDetails:
+		return v.ProtectedTimestampRecord
+	default:
+		panic("not supported")
+	}
+}
+
+// NewManager creates a new protected timestamp provider
+// for a job.
+func NewManager(
+	db *kv.DB,
+	codec keys.SQLCodec,
+	protectedTSProvider protectedts.Provider,
+	systemConfig config.SystemConfigProvider,
+	jr *jobs.Registry,
+) *Manager {
+	return &Manager{
+		db:                  db,
+		codec:               codec,
+		protectedTSProvider: protectedTSProvider,
+		systemConfig:        systemConfig,
+		jr:                  jr,
+	}
+}
+
+// ProtectBeforeGC adds a protected timestamp record for a historical
+// transaction for a specific table, once a certain percentage of the GC time has
+// elapsed. Returns a Cleaner function to cancel installation or remove the
+// protected  timestamp.
+func (p *Manager) ProtectBeforeGC(
```

Comment:
Is this function  supposed to guarantee that it will issue a PTS before the next GC job on the table? Either in this function or in the commit message, could you explain the purpose of this function? I.e. why would the caller use this as oppose to the Protect function?

Reply:
I clarified this with comments on ProtectBeforeGC:
//This method can be preferred in scenarios where the cost of installing
// a protected timestamp is more expensive relative to the typical length of an
// operation (for example multi-region)

and in the commit message:

A new jobsprotectedts.Manager object is introduced, which
can install timestamps synchronous or before the GC TTL
occurs. The asynchronous version ProtectedBeforeGC is
suited for validation where the cost of the protected
timestamp validation could be costly relative to validating
indexes. Whereas, the synchronous version is better suited
for cases where we know its an expensive operation and
a protected timestamp will always be needed.

Reply:
Thanks, this makes thing clearer! One last q, that I think could be clarified. 

This function does not guarantee that a PTS will be written before the next gc run, right? It only guarantees that a PT will be issued in `time.Duration(float64(time.Duration(zoneCfg.GC.TTLSeconds)*time.Second) *
			timedProtectTimeStampGCPct)` seconds. I.e. if a GC is about to run, the PTS will get issued after this next GC job. IIUC, then perhaps this function should be renamed.

Reply:
That's a good point and both of them have the property to update existing records inside them which needs to be documented too. I'm tempted to call it TryToProtectBeforeGC, and document that it's best effort bases. For validation, this is done early so it should be pretty effective unless the GC TTL is crazy small.

Reply:
Added for the TryToProtectBeforeGC:
// The approach here is taken is heuristic
// and can be considered a best-effort basis since the GC TTL could change or
// the caller may not invoke this early enough in the transaction.

Added for Protect:
// If an existing record is found, it will be updated with
// a new timestamp

Reply:
sweet! this looks good to me, though I'll wait for andrew to give this a stamp. The DR refactors look good too.